### DISCUSSION TITLE
3rd Party jobs and `'queues_are_tenant_aware_by_default' => true`

### DISCUSSION THREAD

**Discussion by telkins** [spatie/laravel-multitenancy#235] [Q&A]
Hi,

In one of my apps, I believe every job is or should be tenant aware, so it makes good sense to set `'queues_are_tenant_aware_by_default' => true` in the config file.

But...

I've just brought in another package that has some job that I can't directly modify and I'm getting the `CurrentTenantCouldNotBeDeterminedInTenantAwareJob` exception.

Some more detail: I'm running an artisan command that eventually sends out mails and the `Mailable` that is to be sent implements `ShouldQueue`....but it's not explicitly implementing `NotTenantAware` and I don't have access to that class, either.

Is there a way to keep all queues tenant aware by default and somehow suppress it when running `artisan`...?

I'm not even sure what all the different possibilities are.  I can think of a couple, but neither of them are fun/pretty.  I'm hoping some of the smart people will chime in with some good suggestions. 🤓

Let me know if I can answer any questions to make this easier.

Thanks. 😄

**Comment by masterix21** [comment]
Should be necessary to add a bool config "throw an error if job arent tenant aware" with the default value to `true`: when it is `false` your code won't throw the exception. Is it what you need, right?

**Reply by telkins** [reply]
Perhaps, if I understand you correctly.

It sounds like you're suggesting adding a new config, `throw_error_if_job_isn't_tenant_aware`, defaulting to `true`.  With this setting, everything would be like it is today.

But, if it's set to `false`, then when a job is encountered that is neither `TenantAware` nor `NotTenantAware`, no exception would be thrown...and the job would be executed, anyway.

Please confirm if this is what you're suggesting.

If this is what you're suggesting, the problem is that the job still needs to be processed for the landlord or for the current tenant.  For my particular needs, the 3rd-party job should be considered `NotTenantAware`.  But, I'd guess that other scenarios would required different handling.

So, off the top of my head, perhaps there should be **two** new config settings, both arrays...something like this:
```php
    'tenant_aware_jobs' => [
        ThirdParty\Jobs\DoSomething::class,
    ],
    'not_tenant_aware_jobs' => [
        ThirdParty\Jobs\DoSomethingElse::class,
    ],
```

Then, whenever a job's discovered where an exception would normally be thrown, these lists can be considered and, if found, then the job can be handled accordingly.

Not sure if this makes sense or if it's doable, but thinking/talking through this leads me to thinking that this could work.

Thoughts...? 🤔

**Reply by masterix21** [reply]
Your idea to use the arrays is ok for me. I will work on it.

**Reply by masterix21** [reply]
@freekmurze what's your idea about it? Can I work for a PR?

**Reply by freekmurze** [reply]
Yup, PR it! 👍

**Comment by masterix21** [comment]
PR released :) sorry for the delay, but I was busy.

**Reply by telkins** [reply]
Fantastic...! 🚀  I'll take a look....

Can you link to the PR if/when you have a chance?  I'll do it if you don't beat me to it.  Not much time at the moment. 😅

**Reply by telkins** [reply]
Nevermind...found it: #301 🤓
